DOC-926 Added support for tool calls - ollama_chat (#147)
redpanda-data/rp-connect-docs: modules/components/pages/processors/ollama_chat.adoc

include::components:partial$ollama_warning.adoc[]

> Generates responses to messages in a chat conversation, using the Ollama API and external tools.

Reviewer: ```suggestion
Generates responses to messages in a chat conversation using the Ollama API and external tools.
```

Author: Updated.

--
======

> This processor sends prompts to your chosen Ollama large language model (LLM) and generates text from the responses, using the Ollama API and external tools.

Reviewer: ```suggestion
This processor sends prompts to your chosen Ollama large language model (LLM) and generates text from the responses. It uses the Ollama API and external tools.
```

> By default, the processor starts and runs a locally installed Ollama server. Alternatively, to use an already running Ollama server, add your server details to the `server_address` field. You can https://ollama.com/download[download and install Ollama from the Ollama website^].

Reviewer: ```suggestion
By default, the processor starts and runs a locally-installed Ollama server. Alternatively, to use an already running Ollama server, add your server details to the `server_address` field. You can https://ollama.com/download[download and install Ollama from the Ollama website^].
```

Author: Updated.

The prompt you want to generate a response for. By default, the processor submits the entire payload as a string.
This field supports xref:configuration:interpolation.adoc#bloblang-queries[interpolation functions].

*Type*: `string`

=== `system_prompt`

The system prompt to submit to the Ollama LLM.
This field supports xref:configuration:interpolation.adoc#bloblang-queries[interpolation functions].

*Type*: `string`

=== `image`

An optional image to submit along with the <<prompt, `prompt`>> value. The result is a byte array.

> *Type*: `string`

Reviewer 2: the type should be bloblang...

=== `stop`

Sets the stop sequences to use. When this pattern is encountered, the LLM stops generating text and returns the final response.

*Type*: `array`

> === `save_prompt_metadata`

Author: I've assumed that all new fields should also be available in Cloud. Is this correct?

Reviewer 2: yes

=== `stop`

Sets the stop sequences to use. When this pattern is encountered, the LLM stops generating text and returns the final response.

*Type*: `array`

=== `save_prompt_metadata`

Set to `true` to save the prompt value to a metadata field (`@prompt`) on the corresponding output message. If you use the `system_prompt` field, its value is also saved to an `@system_prompt` metadata field on each output message.

*Type*: `bool`

*Default*: `false`

=== `max_tool_calls`

The maximum number of sequential calls you can make to external tools to retrieve additional information to answer a prompt.

*Type*: `int`

*Default*: `3`

=== `tools`

> The external tools the LLM can invoke, such as functions, APIs, or web browsing. You can build sub-pipelines that include definitions of these tools, and the LLM chooses when to invoke them to help answer a prompt. For more details, see <<Examples, examples>>.

Reviewer: ```suggestion
The external tools the LLM can invoke, such as functions, APIs, or web browsing. You can build subpipelines that include definitions of these tools, and the LLM chooses when to invoke them to help answer a prompt. For more information, see <<Examples, examples>>.
```
@asimms41 just my preference

Author: I looked it up and I couldn't find subpipelines in dictionaries.

Author: Updated, replacing term subpipelines.

*Type*: `string`

=== `tools[].name`

The name of the external tool you want to use.

*Type*: `string`

=== `tools[].description`

> A description of what the tool does, which the LLM uses to decide when to invoke it.

Reviewer: ```suggestion
A description of what the tool does. The LLM uses this to decide when to invoke the tool.
```

Author: Updated.

*Type*: `string`

== Examples

[tabs]
=====
Analyze an image and generate a description for it::
+
--
This configuration fetches image URLs from stdin and uses the LLaVA LLM to describe the images.

```yaml
input:
  stdin:
    scanner:
      lines: {}
pipeline:
  processors:
    - http:
        verb: GET
        url: "${!content().string()}"
    - ollama_chat:
        model: llava
        prompt: "Describe the following image"
        image: "root = content()"
output:
  stdout:
    codec: lines
```
--

> Use a sub-pipeline to make calls to external tools::

Reviewer: ```suggestion
Use a subpipeline to make calls to external tools::
```

Author: Updated.

+
--

> In this configuration, the LLama 3.2 model executes a sub-pipeline, which makes a tool call to retrieve weather data for a specific city.

Reviewer: ```suggestion
In this configuration, the LLama 3.2 model executes a subpipeline, which makes a tool call to retrieve weather data for a specific city.
```

Author: See previous comment.

Author: Updated.
